meta-mender-tegra/recipes-bsp/u-boot/patches/0001-Move-jetson-nano-env-to-SDcard.patch

```diff
@@ -1,41 +1,56 @@
-From 0ca8cf728807766f9bc36fe8551b96f3295273b5 Mon Sep 17 00:00:00 2001
+From 5c80fcad4f8b1c04b246fd3315dbd0fcc68d471d Mon Sep 17 00:00:00 2001
 From: Matt Madison <matt@madison.systems>
 Date: Fri, 10 Jan 2020 12:19:14 -0800
-Subject: [PATCH] Move jetson-nano env to SDcard
+Subject: [PATCH 8/8] Move jetson-nano env to SDcard
```

Comment:
I think the title and filename on this patch is wrong now?  Looks like it's keeping environment in emmc?
I don't completely have my head wrapped around how these changes interact with `TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH`.  
Also we haven't tested this configuration yet with the nano dev kit and SDcard yet but need to do that if we plan to keep supporting `TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH`.

Reply:
This patch is still only applied for t210 when `TEGRA_SPIFLASH_BOOT == 1` and `TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH != 1`, as it was before https://github.com/OE4T/meta-mender-community/blob/15635b9d9ed8d9064d32a5e3763b355520e54359/meta-mender-tegra/recipes-bsp/u-boot/u-boot-mender-tegra.inc#L7
I did my best to port this since there were a few changes in this area, but I don't have a nano dev kit to test this out on, so a sanity check that this is still working would be good.

Reply:
> This patch is still only applied for t210 when TEGRA_SPIFLASH_BOOT == 1 and TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH != 1, as it was before

Got it, I was confused by the `CONFIG_ENV_IS_IN_MMC=y` but I think MMC here doesn't refer to emmc it refers to the SDcard and I had forgotten this.

> but I don't have a nano dev kit to test this out on, so a sanity check that this is still working would be good.

We'll take care of this, we plan to test on these platforms:
 * jetson-nano-emmc-qspi-sd (which should test SDcard boot in default configuration)
 * jetson-tx2
 * jetson-tx2 with PREFERRED_PROVIDER_virtual/bootloader ?= "cboot-prebuilt" set as described in https://github.com/madisongh/meta-tegra/wiki/Using-cboot-as-bootloader,  
 * jetson-xavier-nx-devkit-emmc
 * jetson-nano-emmc

Reply:
Unfortunately we are seeing a problem with `jetson-nano-emmc-qspi-sd`.  It looks like uboot configuration isn't working properly.  We can't read environment successfully after setting it from `fw_printenv`.  We are going to debug this with a serial port connection.

Reply:
@dwalkes Is your meta-tegra up-to-date with dunfell-l4t-r32.4.3? I discovered and fixed a mismatch between the fw_env.config file and the compiled-in environment offset for the nano-qspi-sd last week.

Reply:
> @dwalkes Is your meta-tegra up-to-date with dunfell-l4t-r32.4.3?

We were using the latest master, so we should have commit madisongh/meta-tegra@0d6c1fb I believe. I've updated the full layer setup we are testing at https://github.com/BoulderAI/test-distro/tree/master%2Btegra-updates/repos

Reply:
We are still looking at this.  Writes to uboot environment from the uboot prompt work fine and we can see the changes reflected in userspace, so the problem is the write from userspace via `fw_setenv`.  It looks like `fw_setenv` isn't actually writing anything based on a glance at the strace output.  We are going to compare behavior on TX2 uboot next.

Reply:
I took a quick look, and I see the regular fw_env.config file from the meta-tegra layer installed in /etc rather than the version with the Mender modifications.  The Mender one gets generated as fw_env.config.default during the u-boot build, and gets deployed with that name, but something needs to replace the normal one that gets installed by the u-boot-tegra recipe with the generated one.  Not sure where that's supposed to happen.

Reply:
The `/etc/fw_env.config` that is installed comes from libubootenv, and is a symlink to `/data/u-boot/fw_env.config` -- this is done by a bbappend for libubootenv (./meta-mender-core/recipes-bsp/u-boot/libubootenv_%.bbappend) that first looks for `${DEPLOY_DIR_IMAGE}/fw_env.config.default`, and creates it if it can't find it. It does create the bitbake dependency on `u-boot:do_deploy`, so it should be picking that up.

That being said, if it can read the variables written by u-boot, it must be configured (at least in part) sanely?

Reply:
> The /etc/fw_env.config that is installed comes from libubootenv

This is confusing me then.  The u-boot recipe packages up `/etc/fw_env.config` in `u-boot-env` (or `u-boot-tegra-env` in our case), and that should get installed in the rootfs... it does for the normal setup, anyway.

> That being said, if it can read the variables written by u-boot, it must be configured (at least in part) sanely?

The content of the generated file looked right to me.   If that's really the copy of the file that's getting used, then it should be aligned with what U-Boot itself is using.

Reply:
> This is confusing me then. The u-boot recipe packages up /etc/fw_env.config in u-boot-env (or u-boot-tegra-env in our case), and that should get installed in the rootfs... it does for the normal setup, anyway.

And how is this package normally installed in the image? I see that the -env recipe is an RDEPENDS on u-boot, but u-boot isn't installed, and nothing else I can find is asking for -env, otherwise there would be a conflict. On a core-image-base install, I've got this:

```
root@jetson-tx2:~# ls -l /etc/fw_env.config 
lrwxrwxrwx    1 root     root            26 Mar  9  2018 /etc/fw_env.config -> /data/u-boot/fw_env.config
root@jetson-tx2:~# opkg search /etc/fw_env.config                           
libubootenv0.2 - 0.2+git0+86bd30a14e-r0
root@jetson-tx2:~# opkg search /data/u-boot/fw_env.config
libubootenv0.2 - 0.2+git0+86bd30a14e-r0
```

Reply:
> We are going to compare behavior on TX2 uboot next.

TX2 uboot and `jetson-nano-emmc` both work fine, so this appears to be something about libubootenv support for [mtdblk device write](https://github.com/sbabic/libubootenv/blob/4a0a4662e25471e8fcaf8767a4fa4f454c88a2e2/src/uboot_env.c#L610) and/or the configuration for mtdblk devices.  

strace on `jetson-nano-emmc-qspi-sd`
```
openat(AT_FDCWD, "/var/lock/fw_printenv.lock", O_WRONLY|O_CREAT|O_TRUNC, 0666) = 3
flock(3, LOCK_EX)                       = 0
mmap(NULL, 266240, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f900be000
openat(AT_FDCWD, "/dev/mtdblock0", O_RDONLY) = 4
lseek(4, 3866624, SEEK_SET)             = 3866624
read(4, "\224\305\35\221\1altbootcmd=run mender_altbo"..., 131072) = 131072
close(4)                                = 0
openat(AT_FDCWD, "/dev/mtdblock0", O_RDONLY) = 4
lseek(4, 3997696, SEEK_SET)             = 3997696
read(4, "\227\350\254C\0altbootcmd=run mender_altbo"..., 131072) = 131072
close(4)                                = 0
munmap(0x7f900be000, 266240)            = 0
brk(0x556dd03000)                       = 0x556dd03000
openat(AT_FDCWD, "/dev/mtdblock0", O_RDWR) = 4
close(4)                                = 0
flock(3, LOCK_UN)                       = 0
close(3)                                = 0
exit_group(0)                           = ?
```
shows reads but no writes.  The strace on  `jetson-nano-emmc` shows the write as expected
```
openat(AT_FDCWD, "/var/lock/fw_printenv.lock", O_WRONLY|O_CREAT|O_TRUNC, 0666) = 3
flock(3, LOCK_EX)                       = 0
mmap(NULL, 266240, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fa0d49000
openat(AT_FDCWD, "/dev/mmcblk0boot1", O_RDONLY) = 4
lseek(4, 3911680, SEEK_SET)             = 3911680
read(4, "4\26\303\300\4altbootcmd=run mender_altbo"..., 131072) = 131072
close(4)                                = 0
openat(AT_FDCWD, "/dev/mmcblk0boot1", O_RDONLY) = 4
lseek(4, 4042752, SEEK_SET)             = 4042752
read(4, "\341\271\247\370\3altbootcmd=run mender_altbo"..., 131072) = 131072
close(4)                                = 0
munmap(0x7fa0d49000, 266240)            = 0
brk(0x55aa4df000)                       = 0x55aa4df000
openat(AT_FDCWD, "/dev/mmcblk0boot1", O_RDWR) = 4
faccessat(AT_FDCWD, "/sys/class/block/mmcblk0boot1/force_ro", W_OK) = 0
openat(AT_FDCWD, "/sys/class/block/mmcblk0boot1/force_ro", O_RDWR) = 5
write(5, "0", 1)                        = 1
close(5)                                = 0
lseek(4, 4042752, SEEK_SET)             = 4042752
write(4, "\341\271\247\370\5altbootcmd=run mender_altbo"..., 131072) = 131072
faccessat(AT_FDCWD, "/sys/class/block/mmcblk0boot1/force_ro", W_OK) = 0
openat(AT_FDCWD, "/sys/class/block/mmcblk0boot1/force_ro", O_RDWR) = 5
fsync(4)                                = 0
write(5, "1", 1)                        = 1
close(5)                                = 0
close(4)                                = 0
```
Nothing's jumping out at me on code review so I think we are going to try setting up gdb.

Reply:
We are still fighting the `jetson-nano-emmc-qspi-sd` machine.  Neither the default case nor `TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH=""` case is working.  We've noticed that in the `TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH=""` case we unexpectedly see the same mtdblk contents in fw_env.config
```
/dev/mtdblock0 0x3b0000 0x20000
/dev/mtdblock0 0x3d0000 0x20000
```
The logic [here](https://github.com/OE4T/meta-mender-community/blob/15635b9d9ed8d9064d32a5e3763b355520e54359/meta-mender-tegra/recipes-bsp/u-boot/u-boot-mender-tegra-vars.inc#L16) is setting `MENDER_UBOOT_MMC_ENV_LINUX_DEVICE_PATH` correctly to `/dev/mmcblk0` but somehow this didn't get refleted in the content of `/etc/fw_env.config`. 
Even if we set this content:
```

Reply:
You'd want those offsets to match the location of the ENV partition on the SDcard.  The offsets you have there are specific to the eMMC-based devices.

Reply:
Thanks @madisongh .  So we'll try hand editing our `/etc/fw_env.config` to
```
/dev/mmcblk0 0xC00000 0x20000
/dev/mmcblk0 0xC20000 0x20000
```
for the case when `TEGRA_MENDER_UBOOT_ENV_IN_SPIFLASH=""` and `TEGRA_SPIFLASH_BOOT=1` which would [match the case for jetson-nano-qspi-sd](https://github.com/madisongh/meta-tegra/blob/master/conf/machine/jetson-nano-qspi-sd.conf#L35).  I suspect this isn't going to do the right thing until we fix the partition layout XML as you've mentioned at https://github.com/OE4T/meta-mender-community/pull/1/files#r467458612

Reply:
There may be some relevant traffic regarding issues with libuboot reading device types in the thread at https://groups.google.com/g/swupdate/c/yoJx3i-Tsmo/m/rgjE_uzzCQAJ as well.

Reply:
I've committed a669d24505633e168c92c4d1db3b64fba76f286a to use /dev/mtd0 instead of /dev/mtdblock0. Please let me know if this fixes up the env in nand flash case.

Reply:
There was a typo in the previous commit that has been rewritten as 031d36e